### PR TITLE
SLEEP-1458 Add instance-like filters on entities

### PR DESCRIPTION
**Pull request by bramj**
I'm really excited about this one!

This PR adds JSON logic filters to the entities page to allow searches such as 'give me all entities that have a form instance with "A" == "100"'.

It uses the `react-awesome-query-builder` in the frontend, formatting the data as "subfields", and adds support in the backend to turn this into SQL `EXISTS` subqueries.

PR on `bluesquare-components`: https://github.com/BLSQ/bluesquare-components/pull/164

Related JIRA tickets : https://bluesquare.atlassian.net/browse/SLEEP-1458

**Note: I'm fairly confident that the generated queries are correct. My main concern is performance when the instances/entities tables grow. I plan to use Trypelim as the guinea pig for this. With about 100K entities and 200K instances it should give us an idea of how this will scale.**

## Self proofreading checklist

- [x] Did I use eslint and black formatters
- [x] Is my code clear enough and well documented
- [x] Are my typescript files well typed
- [x] New translations have been added or updated if new strings have been introduced in the frontend
- ~My migrations file are include~
- [x] Are there enough tests
- ~Documentation has been included (for new feature)~

## Print screen / video

https://github.com/user-attachments/assets/311c8ad4-db7c-4767-a15d-6666ad4d332f

## Notes

Things that the reviewers should know: known bugs that are out of the scope of the PR, other trade-offs that were made.
If the PR depends on a PR in bluesquare-components, or merges into another PR (i.o. main), it should also be mentioned here

### CODE REVIEW

File: hat/assets/js/apps/Iaso/domains/entities/entityTypes/hooks/requests/forms.ts [hunk 1]
```diff
@@ -28,11 +28,15 @@ export const useGetForm = (
 
 export const useGetForms = (
     enabled: boolean,
+    fields?: string[] | undefined,
 ): UseQueryResult<Form[], Error> => {
+    let url = '/api/forms/?fields=id,name,latest_form_version';
+    if (fields) {
+        url += `,${fields.join(',')}`;
+    }
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/domains/entities/entityTypes/hooks/requests/forms.ts]: nitpick but I'm not a big fan of this pattern
```suggestion
    const apiUrl = '/api/forms/?fields=id,name,latest_form_version'; // I would even hoist that outside of the hook
    const url = fields ? `${apiUrl},${fields.join(',')}` : apiUrl;
```

File: hat/assets/js/apps/Iaso/domains/entities/entityTypes/hooks/requests/forms.ts [hunk 2]
```diff
@@ -28,11 +28,15 @@ export const useGetForm = (
...
     return useSnackQuery({
         queryKey: ['forms'],
-        queryFn: () =>
-            getRequest('/api/forms/?fields=id,name,latest_form_version'),
+        queryFn: () => getRequest(url),
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/domains/entities/entityTypes/hooks/requests/forms.ts:44]: If you merge the previous suggestion:

```suggestion
        queryFn: () => getRequest(apiUrl),
```

File: hat/assets/js/apps/Iaso/domains/forms/fields/hooks/useGetFormDescriptor.ts [hunk 3]
```diff
@@ -12,7 +12,7 @@ type FormVersionsList = {
 };
 
 const getVersion = (formId: number | undefined): Promise<FormVersionsList> => {
-    return getRequest(`/api/formversions/?form_id=${formId}&fields=descriptor`);
+    return getRequest(`/api/formversions/?form_id=&fields=descriptor`);
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/domains/forms/fields/hooks/useGetFormDescriptor.ts]: If you need to skip the form id for the feature to work, I would split the code or add a parameter to `useGetFormDescriptor` as it needlessly increases the payload of the request for the submission/instances page (where the form id is known).

File: hat/assets/js/apps/Iaso/domains/forms/fields/hooks/useGetFormDescriptor.ts [hunk 4]
```diff
@@ -25,7 +25,7 @@ export const useGetFormDescriptor = (
         queryKey,
         queryFn: () => getVersion(formId),
         options: {
-            enabled: Boolean(formId),
+            // enabled: Boolean(formId), TODO
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/domains/forms/fields/hooks/useGetFormDescriptor.ts]: Same as above: removing this duplicates the API call on the submissions page

File: hat/assets/js/apps/Iaso/domains/forms/hooks/useGetPossibleFields.ts [hunk 5]
```diff
@@ -41,6 +53,40 @@ export const useGetPossibleFields = (formId?: number): Result => {
...
+                name: form.name,
+                possibleFields,
+            });
+        });
+
+        return {
+            allPossibleFields,
+            isFetchingForms,
+        };
+    }, [isFetchingForms]);
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/domains/forms/hooks/useGetPossibleFields.ts]:
```suggestion
    }, [isFetchingForms, allForms]);
```
You need to update the return value if `allForms` change. Normally your linter should warn you if your deps array is incomplete.

File: hat/assets/js/apps/Iaso/domains/forms/hooks/useGetPossibleFields.ts [hunk 6]
```diff
@@ -41,6 +53,40 @@ export const useGetPossibleFields = (formId?: number): Result => {
     return usePossibleFields(isFetchingForm, currentForm);
 };
 
+export const useAllPossibleFields = (
+    isFetchingForms: boolean,
+    allForms: Form[] = [],
+): AllResults => {
+    return useMemo(() => {
+        let allPossibleFields: PossibleFieldsForForm[] = [];
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/domains/forms/hooks/useGetPossibleFields.ts]:
```suggestion
        const allPossibleFields: PossibleFieldsForForm[] = [];
```
Pushing values into the array doesn't change its reference, so its safe (and preferred) to declare it with `const`

File: hat/assets/js/apps/Iaso/domains/forms/hooks/useGetPossibleFields.ts (hunk 5 above)
Comment (quang-le) [reply hat/assets/js/apps/Iaso/domains/forms/hooks/useGetPossibleFields.ts]: If done on purpose, please leave a comment explaining the reason why, and why it's safe to do so

**Review by quang-le — CHANGES_REQUESTED** [review]
Some code style changes + 2 blocking:
- Please adapt `useGetFormDescriptor` to avoid double fetch + fetching all forms on instances page
- Dependencies array of `useMemo` should not be incomplete

File: iaso/utils/jsonlogic.py [hunk 7]
```diff
@@ -36,28 +36,34 @@ def as_sql(self, compiler, connection):
...
     """
 
+    func = jsonlogic_to_q if recursion_func is None else recursion_func
+
     if "and" in jsonlogic:
         return reduce(
             operator.and_,
-            (jsonlogic_to_q(subquery, field_prefix) for subquery in jsonlogic["and"]),
+            (func(subquery, field_prefix) for subquery in jsonlogic["and"]),
         )
```
Comment (kemar) [inline iaso/utils/jsonlogic.py]: I know that was implemented before and this is largely a personal preference but… every time I see a `reduce()` call, I need to go read the doc to understand what's actually happening.

I tend to think that an accumulation loop is easier to understand, e.g.

```python
sub_query = Q()
for lookup in jsonlogic["and"]:
    sub_query = operator.and_(sub_query, func(lookup, field_prefix))
return sub_query
```

File: iaso/tests/api/test_entities.py [hunk 8]
```diff
@@ -421,12 +525,12 @@ def test_retrieve_entity_only_same_account(self):
...
-            reference_form=self.form_1,
-            fields_detail_info_view=["something", "else"],
-            fields_list_view=["something", "else"],
-        )
+        # entity_type = EntityType.objects.create(
+        #     name="Type 1",
+        #     reference_form=self.form_1,
+        #     fields_detail_info_view=["something", "else"],
+        #     fields_list_view=["something", "else"],
+        # )
```
Comment (kemar) [inline iaso/tests/api/test_entities.py:533]: Maybe this could be deleted then?

**Review by kemar — COMMENTED** [review]
Backend LGTM

Kudos because this is advanced work with the ORM.

I'm also concerned with performance and curious to hear about how Trypelim's going to handle it.

File: hat/assets/js/apps/Iaso/domains/forms/fields/hooks/useGetFormDescriptor.ts (hunk 3 above)
Comment (bramj) [reply hat/assets/js/apps/Iaso/domains/forms/fields/hooks/useGetFormDescriptor.ts]: Yes absolutely, it needs to be a separate hook, totally forgot this TODO

File: hat/assets/js/apps/Iaso/domains/entities/entityTypes/hooks/requests/forms.ts (hunk 1 above)
Comment (bramj) [reply hat/assets/js/apps/Iaso/domains/entities/entityTypes/hooks/requests/forms.ts]: Not sure I find it more readable, but changed :+1:

File: hat/assets/js/apps/Iaso/domains/forms/hooks/useGetPossibleFields.ts (hunk 5 above)
Comment (bramj) [reply hat/assets/js/apps/Iaso/domains/forms/hooks/useGetPossibleFields.ts]: Hm good catch, my linter didn't tell me anything about this, should it come from the `show-lint-problems` task?

File: hat/assets/js/apps/Iaso/domains/entities/entityTypes/hooks/requests/forms.ts (hunk 2 above)
Comment (bramj) [reply hat/assets/js/apps/Iaso/domains/entities/entityTypes/hooks/requests/forms.ts:44]: I've made the changes you proposed :+1: there it's the `url` we want though

**Comment by bramj** [comment]
> Backend LGTM
> 
> Kudos because this is advanced work with the ORM.
> 
> I'm also concerned with performance and curious to hear about how Trypelim's going to handle it.

So far so good! But it's only on about 100K beneficiaries for now. The good part about the QueryBuilder is that the frontend is already prepared for if/when we want to use something like Elastic search to return the results from the backend.

File: hat/assets/js/apps/Iaso/domains/forms/fields/hooks/useGetQueryBuildersFields.ts [hunk 9]
```diff
@@ -54,3 +55,30 @@ export const useGetQueryBuildersFields = (
...
+        fields[form_id] = {
+            label: name,
+            type: '!group',
+            mode: 'array',
+            conjunctions: ['AND', 'OR'],
+            operators: ['some', 'all', 'none'],
+            defaultOperator: 'some',
+            subfields: subfields,
+        };
+    }
```
Comment (beygorghor) [inline hat/assets/js/apps/Iaso/domains/forms/fields/hooks/useGetQueryBuildersFields.ts:81]:
```suggestion
export const useGetQueryBuilderFieldsForAllForms = (
    formDescriptors?: FormDescriptor[],
    allPossibleFields?: PossibleFieldsForForm[],
): QueryBuilderFields => {
    if (!allPossibleFields || !formDescriptors) return {};
    
    return allPossibleFields.reduce((fields, { form_id, name, possibleFields }) => {
        fields[form_id] = {
            label: name,
            type: '!group',
            mode: 'array',
            conjunctions: ['AND', 'OR'],
            operators: ['some', 'all', 'none'],
            defaultOperator: 'some',
            subfields: useGetQueryBuildersFields(formDescriptors, possibleFields),
        };
        return fields;
    }, {} as QueryBuilderFields);
};
```

Comment (beygorghor) [reply hat/assets/js/apps/Iaso/domains/forms/fields/hooks/useGetQueryBuildersFields.ts:81]: not a big fan of iterators

File: hat/assets/js/apps/Iaso/domains/forms/hooks/useGetPossibleFields.ts [hunk 10]
```diff
@@ -42,6 +57,40 @@ export const useGetPossibleFields = (formId?: number, appId?: string): Result =>
...
+                possibleFields,
+            });
+        });
+
+        return {
+            allPossibleFields,
+            isFetchingForms,
+        };
+    }, [isFetchingForms, allForms]);
+};
```
Comment (beygorghor) [inline hat/assets/js/apps/Iaso/domains/forms/hooks/useGetPossibleFields.ts:84]: 😮  Do you really get all possible fields for all forms ? this is a huge query to the backend and can be a big payload. This is one reason why we have to preselect a form to enable querybuilder on instances page

**Review by beygorghor — APPROVED** [review]
Small remarks, i'm just wondering if the call to get all forms and their possible field won't crash at some point, do we use some cache in this ?

**Comment by beygorghor** [comment]
can you please move those fields 
Screenshot 2024-09-05 at 14 24 51

File: hat/assets/js/apps/Iaso/domains/forms/hooks/useGetPossibleFields.ts (hunk 5 above)
Comment (quang-le) [reply hat/assets/js/apps/Iaso/domains/forms/hooks/useGetPossibleFields.ts]: It's from `react-hooks/exhaustive-deps`

**Review by quang-le — APPROVED** [review]

**Pull request merged**